src/audio/google/google_ctc_audio_processing.c

```diff
@@ -0,0 +1,359 @@
+						      &google_ctc_audio_processing_config_present);
+
+	if (google_ctc_audio_processing_config_present) {
+		comp_info(dev,
+			  "google_ctc_audio_processing_reconfigure(): Applying config of size %zu bytes",
+			  google_ctc_audio_processing_config_size);
+
+		ret = GoogleCtcAudioProcessingReconfigure(cd->state,
+							  google_ctc_audio_processing_config,
+							  google_ctc_audio_processing_config_size);
+		if (ret) {
+			comp_err(dev, "GoogleCtcAudioProcessingReconfigure failed: %d",
+				 ret);
+			return ret;
+		}
+	}
+
+	return 0;
+}
+
+static int ctc_prepare(struct processing_module *mod)
+{
+	comp_info(mod->dev, "ctc_prepare()");
+
+	return 0;
+}
+
+static int ctc_reset(struct processing_module *mod)
+{
+	comp_dbg(mod->dev, "ctc_reset()");
+
+	return 0;
+}
+
+static int ctc_process(struct processing_module *mod,
+					       struct sof_source **sources, int num_of_sources,
+					       struct sof_sink **sinks, int num_of_sinks)
+{
+	struct google_ctc_audio_processing_comp_data *cd = module_get_private_data(mod);
+
+	int16_t const *in;
+	int8_t const *in_buf_start;
+	int8_t const *in_buf_end;
+	size_t in_buf_size;
+
+	int16_t *out;
+	int8_t *out_buf_start;
+	int8_t *out_buf_end;
+	size_t out_buf_size;
+
+	size_t num_of_bytes_to_process;
+
+	struct sof_source *in_stream;
+	struct sof_sink *out_stream;
+
+	comp_dbg(dev, "ctc_process()");
+
+	if (cd->reconfigure) {
+		ret = google_ctc_audio_processing_reconfigure(mod);
+		if (ret)
+			return ret;
+	}
+
+	in_stream = sources[0];
+	out_stream = sinks[0];
+
+	num_of_bytes_to_process = cd->num_frames * source_get_frame_bytes(in_stream);
+
+	ret = source_get_data(in_stream, num_of_bytes_to_process, (const void **)&in,
+			      (const void **)&in_buf_start,  &in_buf_size);
+	assert(!ret);
+
+	ret = sink_get_buffer(out_stream, num_of_bytes_to_process, (void **)&out,
+			      (void **)&out_buf_start, &out_buf_size);
+	assert(!ret);
+
+	GoogleCtcAudioProcessingProcess(cd->state, in, out, cd->num_frames);
+
+	source_release_data(in_stream, num_of_bytes_to_process);
+	sink_commit_buffer(out_stream, num_of_bytes_to_process);
+
+	return 0;
+}
+
+#if CONFIG_IPC_MAJOR_3
+static int ctc_cmd_set_data(struct processing_module *mod,
+						    struct sof_ipc_ctrl_data *cdata)
+{
+	struct google_ctc_audio_processing_comp_data *cd = module_get_private_data(mod);
+	int ret;
+
+	switch (cdata->cmd) {
+	case SOF_CTRL_CMD_BINARY:
+		ret = comp_data_blob_set_cmd(cd->tuning_handler, cdata);
+		if (ret)
+			return ret;
+		/* Accept the new blob immediately so that userspace can write
+		 * the control in quick succession without error.
+		 * This ensures the last successful control write from userspace
+		 * before prepare/copy is applied.
+		 * The config blob is not referenced after reconfigure() returns
+		 * so it is safe to call comp_get_data_blob here which frees the
+		 * old blob. This assumes cmd() and prepare()/copy() cannot run
+		 * concurrently which is the case when there is no preemption.
```

Comment:
concurrently - on different cores or preempting each other? I'd say, that the only assumption that you can make now is that IPCs are serialised. I.e. a new IPC command cannot be sent before the previous one has been answered. And an IPC is answered only after its processing has been completed by the DSP. The only processes that can run concurrently or preempt each other are the streaming / copying taking place in the background and new IPCs (stop, pause, configure, etc.). Then you have to consider which event can preempt which context...